t/042-includegraph-reset-state.t.cpp

```diff
@@ -0,0 +1,113 @@
+#include "clangmetatool-testconfig.h"
+
+#include <gtest/gtest.h>
+
+#include <clangmetatool/meta_tool_factory.h>
+#include <clangmetatool/meta_tool.h>
+#include <clangmetatool/collectors/include_graph.h>
+#include <clangmetatool/include_graph_dependencies.h>
+
+#include <clang/Frontend/FrontendAction.h>
+#include <clang/Tooling/Core/Replacement.h>
+#include <clang/Tooling/CommonOptionsParser.h>
+#include <clang/Tooling/Tooling.h>
+#include <clang/Tooling/Refactoring.h>
+#include <llvm/Support/CommandLine.h>
+
+class MyTool {
+private:
+  clang::CompilerInstance* ci;
+  clangmetatool::collectors::IncludeGraph graph;
+public:
+  MyTool(clang::CompilerInstance* ci, clang::ast_matchers::MatchFinder *f)
+    :ci(ci), graph(ci, f) {
+  }
+  void postProcessing
+  (std::map<std::string, clang::tooling::Replacements> &replacementsMap) {
+    clangmetatool::collectors::IncludeGraphData *data = graph.getData();
+
+    clangmetatool::types::FileUID header = 0;
+    clangmetatool::types::FileUID cpp = 0;
+    for(auto itr = data->fuid2name.cbegin(); itr != data->fuid2name.cend(); ++itr){
+      if(itr->second == "foo.h"){
+        header = itr->first;
+      }
+      if(itr->second == "foo.cpp"){
+        cpp = itr->first;
+      }
+    }
+    auto edge = std::make_pair(cpp, header);
+    size_t edge_rc = 4; // edge initial reference count
+
+    // validator that validate the edge and its reference count
+    // before and after doing decrementUsageRefCount
+    auto validator = [&](){
+      ASSERT_EQ(data->usage_reference_count.count(edge), 1);
+      ASSERT_EQ(data->usage_reference_count[edge], edge_rc);
+      clangmetatool::IncludeGraphDependencies::decrementUsageRefCount(data, edge);
+      ASSERT_EQ(data->usage_reference_count[edge], edge_rc - 1);
+      clangmetatool::IncludeGraphDependencies::decrementUsageRefCount(data, edge);
+      ASSERT_EQ(data->usage_reference_count[edge], edge_rc - 2);
+    };
+
+    // backup and validate the initial state 
+    clangmetatool::IncludeGraphDependencies::backup(data);
+    validator();
+    
+    // validate that the include graph is able to be restored
+    clangmetatool::IncludeGraphDependencies::restore(data);
```

Comment:
Why not just `data = graph.getData()`?

Reply:
Only `usage_reference_count` in `IncludeGraphData` is supposed to be changed by client and be backup, the rest data structures should be determined at clang compile time

Reply:
So is copying the `usage_reference_count` only an optimization? Is it warranted? How long does the copy of an entire `IncludeGraphData` take?

Reply:
It is an optimization but I think it is more about logic.

I am thinking about the connection between `IncludeGraph`, `IncludeGraphData` and `IncludeGraphDependencies`, open for discussion but here are some of my thoughts:
- once clang frontend action finished, `IncludeGraphData` should be determined and unchanged
  - since it represents a compilation result
  - `IncludeGraph.getData()` returns `IncludeGraphData` by non `const` pointer so it is volatile and can be misused easily
- any component(besides ppcallback, frontend action) that needs to change `IncludeGraphData` should do a copy to maintain the state themselves instead of directly writing to the origin `IncludeGraphData`
  - to avoid causing dirty read in other methods
- `IncludeGraphDependencies` should be a viewer of `IncludeGraphData`

So that `IncludeGraphData` is completely stateless(unchanged after frontend action), and `IncludeGraphDependencies` keep state itself.

Reply:
`IncludeGraphDependencies` has no state, it's just a namespace for a bunch of static functions. From it's documentation:
```
/**
 * Collect stateless functions to query and and modify the state of
 * dependencies of a given `clangmetatool::IncludeGraphData` structure.
 */
```